server/src/main/java/com/google/coffeehouse/servlets/RetrieveTokenServlet.java

```diff
@@ -0,0 +1,156 @@
+  public static final String LOG_INVALID_ID_TOKEN_MESSAGE = 
+      "LOGGING: ID token failed verification: ";
+
+  private static final Gson gson = new Gson();
+  private static final HttpTransport transport = new NetHttpTransport();
+  private static final JacksonFactory jsonFactory = JacksonFactory.getDefaultInstance();
+  private GoogleIdTokenVerifier verifier =
+      new GoogleIdTokenVerifier.Builder(transport, jsonFactory).build();
+  private static final String CLIENT_SECRET = getSecret();
+  private static final String CLIENT_ID =
+      "893627513276-o7p1m433c9l828svolutrbaibqqvmt8q.apps.googleusercontent.com";
+  
+  // Empty code and redirectUri will be overwritten before executing request.
+  private GoogleAuthorizationCodeTokenRequest tokenRequest =
+      new GoogleAuthorizationCodeTokenRequest(
+          transport, jsonFactory, CLIENT_ID, CLIENT_SECRET, /* code= */ "", /* redirectUri= */ "");
+
+  /** 
+   * Overloaded constructor for dependency injection.
+   * @param verifier the class that verifies the validity of the ID token
+   * @param tokenRequest the class that makes the request to exchange auth code for token
+   */
+  public RetrieveTokenServlet(
+      GoogleIdTokenVerifier verifier, GoogleAuthorizationCodeTokenRequest tokenRequest) {
+    super();
+    this.verifier = verifier;
+    this.tokenRequest = tokenRequest;
+  }
+
+  public RetrieveTokenServlet() {
+    super();
+  }
+
+  private static String getSecret() {
+    try {
+      SecretManagerServiceClient client = SecretManagerServiceClient.create();
+      SecretVersionName secretVersionName = SecretVersionName.of(
+            "coffeehouse-step2020", "GoogleOauthClientSecret", "1");
+      AccessSecretVersionResponse response = client.accessSecretVersion(secretVersionName);
+      return response.getPayload().getData().toStringUtf8();
+    } catch (Exception e) {
+      throw new RuntimeException("Unable to get Oauth client secret");
+    }
+  }
+
+  /** 
+   * Exchanges an auth code for an ID/auth/refresh token. Returns the ID token in JSON format.
+   * @param request the POST request that must have a JSON body with a valid auth code and 
+   *     redirect URI (associated with the keys defined by the CODE_KEY_NAME and 
+   *     REDIRECTURI_KEY_NAME constants).
+   * @param response the response from this method, will contain the retrieved ID token as JSON.
+   *     If the request object does not have the required JSON, this object will send a 
+   *     "400 Bad Request error". If the ID token retrieved by this servlet cannot be verified,
+   *     this object will send a "403 Forbidden error"
+   * @throws IOException if an input or output error is detected when the servlet handles the request
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String code;
+    String redirectUri;
+    try {
+      Map body = gson.fromJson(request.getReader(), Map.class);
```

Comment:
I've opted to use the Map approach here because the field names are referenced in the corresponding test files. It would be possible to place these field name declarations in the definition of the new class, but at that point we're not really getting any benefit from the custom deserialization (and we'd still have to check if those fields were populated by the deserialization).